docs/proposals/odo-debug.md

```diff
@@ -0,0 +1,82 @@
+# odo debug support proposal
+
+Table of Contents:
+- [odo debug support proposal](#odo-debug-support-proposal)
+  - [Design overview](#design-overview)
+    - [New odo commands](#new-odo-commands)
+      - [`odo debug start`](#odo-debug-start)
+      - [`odo debug stop`](#odo-debug-stop)
+      - [`odo debug status`](#odo-debug-status)
+    - [`odo debug port-forward LOCAL_PORT:REMOTE_PORT`](#odo-debug-port-forward-localportremoteport)
+- [Languages support](#languages-support)
+  - [NodeJS](#nodejs)
+    - [Problem 1](#problem-1)
+      - [Solution 1](#solution-1)
+      - [Solution 2](#solution-2)
+  - [Java](#java)
+
+## Design overview
+
+The debug mode inside the container is controlled by `DEBUG_MODE` and `DEBUG_PORT` env variable.
+- if `DEBUG_MODE == true` than the code should be executed with remote debugging capabilities enabled.
+- `DEBUG_PORT` is optional and controls the port of the remote debugger.
+
+Odo sets up port-forwarding to allow a debugger to connect to the running process. 
+Debugger client will always connect to `localhost:$DEBUG_PORT`
+### New odo commands
+#### `odo debug start`
+Enables debug mode for the component. It does this by adding `DEBUG_MODE = true` env variable.
+
+After adding env variables the component (Deployment Config) is restarted.
+Once the pod is in running state `odo debug start` executes `odo debug port-forward $DEBUG_PORT:$DEBUG_PORT`.
+This process executed on the background and PID of the process is recorded in `~/.odo/port-forward.pid` file.
+
+
+Flags:
+- `--port`: Optional. Controls value of `DEBUG_PORT` env variable
+- `--context`: Optional. The same as in other commands (see `odo push --context` for example)
+- `--no-port-forward`: Optional. Just enable debugging but don't start forwarding port. User can use `odo debug port-forward` manually or use other mechanisms.
+
+#### `odo debug stop`
+Stop port-forward process if running and remove `DEBUG_PORT` and `DEBUG_MODE` env variables from Deployment Config.
+
+First, terminate and cleanup port-forwarding.
+If `~/.odo/port-forward.pid` exists and process with PID is running terminate the process and remove the file.
+If the process is not running just remove the file (warn the user).
+If the file doesn't exist it is ok, as `odo debug start --no-port-forward` might have been used.
+
+Second, remove `DEBUG_MODE` and `DEBUG_PORT` env variables from Deployment Config and wait for the pod to be the ready state after a restart.
+
+
+#### `odo debug status`
+Informs the user if the debug mode is enabled for the component and if the port-forward process is running or not.
+
+
+### `odo debug port-forward LOCAL_PORT:REMOTE_PORT`
+Forwards `LOCAL_PORT` to container port `REMOTE_PORT`.
+This blocks the terminal. It will be used for `odo debug start`
```

Comment:
Not sure I understand this. When one executes `odo debug port-forward 8080:8080`, it will block the terminal and doing a `Ctrl+C` would stop the port-forwarding. But what does it mean when we say `It will be used for "odo debug start"`. Does it mean that with `odo debug port-forward 8080:8080` done, I can open a different terminal and do `odo debug start` and it will pick-up that localhost's port 8080 is mapped to components 8080 port?

Reply:
The sentence is wrong 😇  It is meant to be "It will be used *in* `odo debug start`".
What I mean by that that the `start` command will actually make use of `port-forward` command. It will execute it as a separate background process and record its pid. `stop` command will then kill `port-forward` process based on recorded pid.

Reply:
I'm still confused, I guess.  `odo debug port-forward LP:RP` won't be used by a user but by `odo debug start`. What if a user executes it on the terminal? Does a user need to use it through a terminal? If no, should we have a separate command `odo debug port-forward LP:RP` or have that logic in `odo debug start` itself?

I could be asking absolutely dumb stuff because I'm confused. :laughing:

Reply:
> I'm still confused, I guess. `odo debug port-forward LP:RP` won't be used by a user but by `odo debug start`.

yes, a normal user won't use `port-forward` command.

>  What if a user executes it on the terminal? Does a user need to use it through a terminal?
 
Doesn't need to, and most likely won't use this command thought terminal. But there might be corner cases where it can be handy to allow it.

>If no, should we have a separate command odo debug port-forward LP:RP or have that logic in odo debug start itself?

`odo debug start` shouldn't block terminal. It will need to either spaw new process or bring itself to the background. Spawning new process seems an easier and more reliable solution to me, especially when we have to deal with windows. 

> I could be asking absolutely dumb stuff because I'm confused. 😆

You are asking the right questions. I should have included more info on why I thought about this design.

I had two use cases in mind:

**standard use** - local and remote ports are the same
this is what I expect will be the most common, 99.9% will use just this one command.
```
odo debug start --port 9090
```

**advance use** - local and remote ports are different (local 9999 forwarded to remote 9090)
this is a little bit more complicated (two commands are required), but most users won't ever need this
```
odo debug start --port 9090 --no-port-forward
odo debug port-forward 9999:9090
```

Reply:
> > What if a user executes it on the terminal? Does a user need to use it through a terminal?
> 
> Doesn't need to, and most likely won't use this command thought terminal. But there might be corner cases where it can be handy to allow it.

Agree with corner use cases point. I can't think of one right away but that's the definition of a corner use-case. :joy:

> > If no, should we have a separate command odo debug port-forward LP:RP or have that logic in odo debug start itself?
> 
> `odo debug start` shouldn't block terminal. It will need to either spaw new process or bring itself to the background. Spawning new process seems an easier and more reliable solution to me, especially when we have to deal with windows.

Not a technical question: And having a unified approach for all platforms would be a good idea for maintainability, right?  Asking for my understanding.

> I had two use cases in mind:
> 
> **standard use** - local and remote ports are the same
> this is what I expect will be the most common, 99.9% will use just this one command.
> 
> ```
> odo debug start --port 9090
> ```

So if a user only does `odo debug start --port 9090`, we will automagically map localhost's port 9090 to container's port 9090 on which the deployed component is listening? I think we should document this default behaviour. I couldn't find it in this proposal. I think it was sort of assumed in this statement of proposal: 
```
Once the pod is in running state `odo debug start` executes `odo debug port-forward $DEBUG_PORT:$DEBUG_PORT`.
```

> **advance use** - local and remote ports are different (local 9999 forwarded to remote 9090)
> this is a little bit more complicated (two commands are required), but most users won't ever need this
> 
> ```
> odo debug start --port 9090 --no-port-forward
> odo debug port-forward 9999:9090
> ```
:+1: